### PR TITLE
feat: add the insideDataZoom 'gestureOnTouchPad' option that can Optimize the Experience on the MAC touchpad.

### PR DESCRIPTION
**Pull request by jianqi-jin**
## Brief Information
This PR focuses on enhancing the experience when we just use the MAC touchpad to move or scale the `dataZoom`.
fix: #17286
This pull request is in the type of:

- [ ] bug fixing
- [x] new feature
- [ ] others

### Before: What was the problem?

We can't use the touchpad to move the dataZoom. We just can scale it. And I wish that we can use the touchpad to move the dataZoom window. Such as lots of software, they can use the touchpad to scroll x direction and y direction both.

You can see the test file in the `test/gesture.html` see the result.

I added an option `gestureOnTouchPad`: 
Users can set `gestureOnTouchPad` to `true`, that means we adjust the MAC touchpad.
```

                  dataZoom: [
                    {
                      type: 'inside',
                      xAxisIndex: [0],
                      start: 1,
                      zoomOnMouseWheel: true,
                      gestureOnTouchPad: true, // this option
                      end: 35
                    },
                    {
                      type: 'inside',
                      yAxisIndex: [0],
                      zoomOnMouseWheel: true,
                      start: 29,
                      gestureOnTouchPad: true, // this option
                      end: 36
                    }
                  ],
```

https://user-images.githubusercontent.com/33452468/176181799-6cc6732b-4e08-4c95-8557-0feb98ba874d.mov

## Document Info

One of the following should be checked.

- [ ] This PR doesn't relate to document changes
- [x] The document should be updated later
- [x] The document changes have been made in apache/echarts-doc#xxx

### ZRender Changes

- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

N.A.

### Merging options

- [ ] Please squash the commits into a single one when merging.

### CODE REVIEW

**Review by Ovilia — CHANGES_REQUESTED** [review]
Please remove the changes of `dist` and `i18n`.

**Comment by JinJianQi** [comment]
> Please remove the changes of `dist` and `i18n`.

Okay, I removed them and fixed some lint error. Could you help me to review it again, and maybe find something inappropriate? @Ovilia

File: src/component/helper/RoamController.ts [hunk 1]
```diff
@@ -242,6 +252,44 @@ class RoamController extends Eventful<{
         // their event both, and the final behavior is determined
         // by event listener themselves.
 
+        if (gestureOnTouchPad) {
+          const WindowScrollSpeedFactor = .3;
```
Comment (Ovilia) [inline src/component/helper/RoamController.ts:256]: We should probably make these parameters configurable in options.

**Review by Ovilia — CHANGES_REQUESTED** [review]
Just to be clear, do we want to provide a feature that:

A) Automatically turns zooming into scroll for touchpad events

or
 
B) Let the developer decide whether the scroll event should be handled to be zooming or scrolling?

If A, then we should find a way to detect touchpad devices in our code and do not provide an extra option like `gestureOnTouchPad`.

If B, then we should not provide an extra option. It should work with:

```ts
zoomOnMouseWheel: false,
moveOnMouseWheel: true
```

But I tested and it seems a little strange because it uses scroll for both directions. @100pah Should we take this as a bug, or expect mouse wheel not include touch pad?

If this is a bug, then should the dataZoom move only on y direction if a mouse rather than a touchpad is used?

If this is not a bug, then we should provide a option called `natureMoveOnMouseWheel`? Any better idea? `gestureOnTouchPad` is not accurate and it does not detect touch pad device now so the naming is a little strange.

File: src/component/helper/RoamController.ts (hunk 1 above)
Comment (JinJianQi) [reply src/component/helper/RoamController.ts:256]: Yeah, Thanks a lot @Ovilia, I would like to add scrolling both directions as a feature. So I think this not a bug. the name `natureMoveOnMouseWheel` is good for me.

And also, I already set `gestureOnTouchPad` as an options adjust with 
```
zoomOnMouseWheel: false,
moveOnMouseWheel: true,
```
now it looks like:
```
zoomOnMouseWheel: false,
moveOnMouseWheel: true,
gestureOnTouchPad: true,
```
Users can set this option use `setOption();`

**Comment by JinJianQi** [comment]
Personally, I think B is better, case touchpad is so different with mouse, and I think we should offer users to decide how to deal with wheel event.

File: src/component/helper/RoamController.ts (hunk 1 above)
Comment (JinJianQi) [reply src/component/helper/RoamController.ts:256]: > We should probably make these parameters configurable in options.

Sorry, I have little confuse,  which kind of options? Is the options of `InsideZoomModel` ?

**Comment by JinJianQi** [comment]
Any thing update here ~

**Comment by Ovilia** [comment]
@100pah Please help review this PR.

**Comment by JinJianQi** [comment]
@100pah, Is there any thing bad?  If so, I would like to fix it. Thanks. :)

**Comment by 100pah** [comment]
@JinJianQi @Ovilia 
This case shows how to use "mousewheel" to scroll single diectory:
https://echarts.apache.org/examples/en/editor.html?c=custom-gantt-flight

> A) Automatically turns zooming into scroll for touchpad events
> or
> B) Let the developer decide whether the scroll event should be handled to be zooming or scrolling?

In my opinion, that should be B, because what "mousewheel" stands for is probably decided by the app designer rather than users behaviors.

Does [moveOnMouseWheel](https://echarts.apache.org/en/option.html#%2Fsearch%2FmoveOnMouseWheel) already match the requirements?

**Comment by JinJianQi** [comment]
As far as I know, the [moveOnMouseWheel](https://echarts.apache.org/en/option.html#%2Fsearch%2FmoveOnMouseWheel) now just can't distinguish between X and Y directions. To put it simply, I just want to move both directions through touchPad and zoom in/out through separating two fingers (when we separate two fingers, it'll trigger wheel event and the [mouseEvent.ctrlKey](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/ctrlKey) will be true).

Btw, when I set `zoomOnMouseWheel` to `ctrl`, I can actually zoom in/out through touchPad, but the speed is too fast.

In my commit, I just change the method of calculating the required offset, and I would like to set `WindowScrollSpeedFactor` as one of the options, so that users can change the speed when wheel.

**Comment by 100pah** [comment]
@JinJianQi @Ovilia 
I got your point. 
And the effect is good when both x and y direction can be controlled by two finger move and zoom by two fingers separating 👍 . 

I have some suggestions about the `option`: as Ovilia mentioned, the option name `gestureOnTouchPad` is not straightforward enough and attached to "touchpad". I think there are three possible solutions:
+ A. Use `natureMoveOnMouseWheel: true` to represents: two finger move to scroll both x and y direction, and two finger separation to zoom in/out. And leave the behavior of the existing `moveOnMouseWheel: true` `zoomOnMouseWheel: true` not changed.
+ B. Use the existing `moveOnMouseWheel: 'all'` to represents: two finger move to scroll both x and y direction, and two finger separation to zoom in/out.  And leave the behavior of the existing `moveOnMouseWheel: true` `zoomOnMouseWheel: true` not changed.
+ C. Use the existing `moveOnMouseWheel: true` to represents: two finger move to scroll both x and y direction, and two finger separation to zoom in/out.  And drop the original behavior of `moveOnMouseWheel: true` (that is, only can scroll on one direction)

BTW, @JinJianQi,  the code style should follow the [style standard](https://echarts.apache.org/en/coding-standard.html), for example, use 4 spaces indent rather than 2 spaces ~

**Comment by JinJianQi** [comment]
@100pah Thanks for your reply :)

From my side, I think both A and B are good, C just changes the original behavior, which means we may need another way to represent the original behavior. 

In addition, for the style, I'll update this PR later ~

**Comment by tihuan** [comment]
Gentle ping 🙏 Was looking for this exact feature as well 😄 Thanks for the PR!

**Comment by Chaoses-Ib** [comment]
This also works on Windows touchpads. The current scaling factor of ECharts is too large for Windows touchpads. The slightest pinch you make on the touchpad will cause the chart to scale dramatically. This PR solves this problem as well.

However, it looks like `zoomOnMouseWheel: true` is broken with `gestureOnTouchPad: true`. When I scroll the wheel, the chart moves instead of zooming. When I set `gestureOnTouchPad: false`, the chart scales as usual with the scroll wheel.

**Comment by masoudmanson** [comment]
Just wanted to drop a friendly message to check for any updates on this PR. The feature that has been added is exactly what we need for our project, and we're eagerly looking forward to its merge.
Great work, by the way @JinJianQi 💯

**Comment by JinJianQi** [comment]
Sorry for the very late reply, and you guys are awesome.

@Chaoses-Ib I haven't found a suitable method to accurately differentiate actions from the trackpad and the mouse. From the wheel event perspective, vertical scrolling with two fingers on the trackpad is exactly the same as using the mouse wheel. So it seems to me that it's not possible to use the trackpad for movement and the mouse wheel for zooming (because, in my view, their wheel events are the same).

@100pah Since this parameter includes both move and zoom, it will definitely affect zoomOnMouseWheel and moveOnMouseWheel. My current approach is that if natureMoveOnMouseWheel is set to true, then ignore zoomOnMouseWheel and moveOnMouseWheel. If the user wants to use zoomOnMouseWheel and moveOnMouseWheel, they need to set natureMoveOnMouseWheel to false.
Can you help me review this again? @100pah

**Comment by jianqi-jin** [comment]
Since this PR has been opened for a long time. I'll close this and create a new one.

**Pull request closed without merging**